daemon/logger/journald/read.go

```diff
@@ -245,7 +251,7 @@ drain:
 	return cursor
 }
 
-func (s *journald) followJournal(logWatcher *logger.LogWatcher, config logger.ReadConfig, j *C.sd_journal, pfd [2]C.int, cursor *C.char) *C.char {
+func (s *journald) followJournal(logWatcher *logger.LogWatcher, config logger.ReadConfig, j *C.sd_journal, pfd [2]C.int, cursor *C.char, untilUnixMicro uint64) *C.char {
```

Comment:
Oh, darn, I had a reply written down, but it was still "pending" yes - after my earlier comment, I did a quick try to use `logger.ReadConfig` here, instead of a new variable;

```patch
diff --git a/daemon/logger/journald/read.go b/daemon/logger/journald/read.go
index 0e8d6a5..d9481b2 100644
--- a/daemon/logger/journald/read.go
+++ b/daemon/logger/journald/read.go
@@ -170,11 +170,16 @@ func (s *journald) Close() error {
 	return nil
 }
 
-func (s *journald) drainJournal(logWatcher *logger.LogWatcher, config logger.ReadConfig, j *C.sd_journal, oldCursor *C.char, untilUnixMicro uint64) *C.char {
+func (s *journald) drainJournal(logWatcher *logger.LogWatcher, config logger.ReadConfig, j *C.sd_journal, oldCursor *C.char) *C.char {
 	var msg, data, cursor *C.char
 	var length C.size_t
 	var stamp C.uint64_t
 	var priority, partial C.int
+	var until uint64
+
+	if !config.Until.IsZero() {
+		until = uint64(config.Until.UnixNano() / 1000)
+	}
 
 	// Walk the journal from here forward until we run out of new entries
 	// or we reach the until value (if provided).
@@ -196,7 +201,7 @@ drain:
 				break
 			}
 			// Break if the timestamp exceeds any provided until flag.
-			if untilUnixMicro != 0 && untilUnixMicro < uint64(stamp) {
+			if until != 0 && until < uint64(stamp) {
 				break
 			}
 
@@ -251,7 +256,7 @@ drain:
 	return cursor
 }
 
-func (s *journald) followJournal(logWatcher *logger.LogWatcher, config logger.ReadConfig, j *C.sd_journal, pfd [2]C.int, cursor *C.char, untilUnixMicro uint64) *C.char {
+func (s *journald) followJournal(logWatcher *logger.LogWatcher, config logger.ReadConfig, j *C.sd_journal, pfd [2]C.int, cursor *C.char) *C.char {
 	s.mu.Lock()
 	s.readers.readers[logWatcher] = logWatcher
 	if s.closed {
@@ -278,7 +283,7 @@ func (s *journald) followJournal(logWatcher *logger.LogWatcher, config logger.Re
 				break
 			}
 
-			cursor = s.drainJournal(logWatcher, config, j, cursor, untilUnixMicro)
+			cursor = s.drainJournal(logWatcher, config, j, cursor)
 
 			if status != 1 {
 				// We were notified to stop
@@ -409,7 +414,7 @@ func (s *journald) readLogs(logWatcher *logger.LogWatcher, config logger.ReadCon
 			return
 		}
 	}
-	cursor = s.drainJournal(logWatcher, config, j, nil, untilUnixMicro)
+	cursor = s.drainJournal(logWatcher, config, j, nil)
 	if config.Follow {
 		// Allocate a descriptor for following the journal, if we'll
 		// need one.  Do it here so that we can report if it fails.
@@ -421,7 +426,7 @@ func (s *journald) readLogs(logWatcher *logger.LogWatcher, config logger.ReadCon
 			if C.pipe(&pipes[0]) == C.int(-1) {
 				logWatcher.Err <- fmt.Errorf("error opening journald close notification pipe")
 			} else {
-				cursor = s.followJournal(logWatcher, config, j, pipes, cursor, untilUnixMicro)
+				cursor = s.followJournal(logWatcher, config, j, pipes, cursor)
 				// Let followJournal handle freeing the journal context
 				// object and closing the channel.
 				following = true
```

Even though it's a bit cleaner, it also cause the conversion to `untilUnixMicro` to be executed in a loop if `followjournal()` is called (which calls `drainjournal()` in a loop.

So yes, it looks like we need to keep the `untilUnixMicro` argument, but (as @cpuguy83 mentioned above), the `config` isn't used at all, so can be removed